### ISSUE TITLE
[feature] comparable lightning bolt projectile

### ISSUE THREAD

**Issue by lynxlynxlynx** [gemrb/gemrb#146]
The projectile was hardcoded in the original. Currently it hits one target and that's it. 

PEF_BOUNCE (bouncing off walls) and PEF_CONTINUE (continue after hitting a target) need to be implemented. Our projectile is in gemrb/unhardcoded/bg2/lightb.pro and already has both bits set, _just_ the code is missing. At least bg1 is also affected, but likely also the iwds, maybe pst.

**Comment by bradallred** [comment]
We also need to edit something to that the spell is targetable on the ground. What might that be?

**Comment by bradallred** [comment]
I have `PEF_CONTINUE` implemented, but now I'm at a loss for how to determine when to expire the projectile. They don't seem to have a time or distance to limit their lifetime so the lightning endlessly bounces around the map :)

**Comment by lynxlynxlynx** [comment]
The targetting stuff is set in the projectile; I'll change it.
The range is set by the spell (the used extended header), so it either needs to be looked up again when needed or saved somewhere on projectile creation.

Note to self: the only difference in the bg2 projectile is that it has half the speed (30 vs 60):
dce630cc438ca3bf4f683f7438651157  ./gemrb/unhardcoded/iwd/lightb.pro
40df7889e1b68c5535baa28f71cd409f  ./gemrb/unhardcoded/bg2/lightb.pro
dce630cc438ca3bf4f683f7438651157  ./gemrb/unhardcoded/iwd2/lightb.pro
dce630cc438ca3bf4f683f7438651157  ./gemrb/unhardcoded/pst/lightb.pro
dce630cc438ca3bf4f683f7438651157  ./gemrb/unhardcoded/how/lightb.pro

**Comment by lynxlynxlynx** [comment]
At least in bg2, you can only target a creature. Which is great, since the spell has the same targetting mode and we'd need a hack otherwise.

**Comment by bradallred** [comment]
> At least in bg2, you can only target a creature. Which is great, since the spell has the same targetting mode and we'd need a hack otherwise.

Interesting, I could have sworn I had memories of being able to target the ground. I also don't remember the projectile being seeking like ours seems to be.

our code for determining "target" is completely inadequate for this spell. The bolt will only damage the original target (or the caster in certain circumstances) and once it damages them once it becomes inert. I'm pretty sure in the original it was possible to get "double strikes" and of course the bolt continuing on would damage anyone it hit.

I thought the "range" on the spell is for the casting range. I would have thought the lifetime was... well more of a time than a distance, because distance requires extra work to handle and I thought durations on spells was pretty common!

**Comment by lynxlynxlynx** [comment]
I agree the projectile shouldn't be seeking, though I haven't checked. And yes, the bolt damages anyone hit, that's part of PEF_CONTINUE. Maybe the line-aeo (eg. agannazar's scorcher) projectile code has something useful, since one payload application definitely won't be enough. 

The "duration" (spells don't have one, only effects) of the spell is instantaneous, since the bolt is created immediately. Edges of the map are often inaccessible the same way as other stuff (searchmap), so I guess that's why they can bounce. But yeah, shouldn't happen.

**Comment by bradallred** [comment]
Ok, we'll have to keep track of the range in `Projectile` I imagine. The descriptions says it changes with caster level, so a lookup would be harder. The non seeking behavior actually makes this problem easier to solve. Instead of targeting the actor, calculate a line between castor and target and then extend that line to the range of the spell. That solves both the lifetime and seeking behavior bugs.

It also looks like bounce code already exists somewhere, so we just need to find that and use it. When it bounces subtract the distance between the origin and bounce position from the remaining range field and crate a new target point with that distance. Repeat until the distance reaches 0. However, I don't see anything in `Projectile` that seems to handle wall collisions. Do you have any insight on that?

I'm not sure the scorcher damage is super useful here, I believe it calculates an instantaneous line at casting time, but we need a smaller line the length of the projectile that we calculate at each step.

**Comment by bradallred** [comment]
Also, is the "range" in the header in pixels? or is there some conversion we have to do, and if so where is that?

**Comment by lynxlynxlynx** [comment]
I don't know the projectile code well, so the only other idea is that it's part of some area limit check somewhere. The scorcher line moves with its target btw, it's not static. The range in the header is in narglebafs that we then convert with an estimated factor (you can use GetSpellDistance).

**Comment by lynxlynxlynx** [comment]
What I see walking around the lighting room in the first area of bg2 + wanding around:
- sometimes the bolt is not oriented properly, but that's likely just a limit of the orientation count + long bam length
- it goes through walls (as noted)
- it rarely hits, though maybe the room shooter is special — my casts hit

**Comment by bradallred** [comment]
I didn't notice any orientation issues in BG2. I don't believe I changed anything with the orientation code either. I only set a target point and the existing code figures out the orientation. It is possible there is a bug there.

I used the same damage application as scorcher, and we probably need to make adjustments. we need to do some testing to determine 2 things:
1. should we be searching the path for more than one node? There is a comment in the code about this. This is for hits along the length of the bolt.
2. should we have a parameter for the search radius? this would be the "width" of the bolt.

feel free to play around with those 2 numbers. that should fix any hit frequency issues.

**Comment by bradallred** [comment]
I'm not sure how we are supposed to calculate the angle to reflect the bolts. My first thought was that `Wall_Polygon`s were the answer, but toggling their drawing in `GameControl` and reading about them in IESDP tells me they are _not_ the way to do this. There doesn't appear to be anything other than the search map to see that we've hit a wall, but that doesn't tell us anything about the shape of the wall.

Do we have any insights to the originals here? The only thing I can think of is a crude approximation by sampling the searchmap to determine a slope.

**Comment by lynxlynxlynx** [comment]
The searchmap may be too coarse for this, but wall groups are polygons, so it should be somewhat simpler, even with the buffering required. Is the problem that there's not enough of them? I wonder if the original really allowed bouncing off of everything. 

@caillean7 @cygarniczka can you test this a bit please? Dungeons are a good choice, eg. chateau irenicus or the shadow temple. In the latter, in the room just before the shadow jailor, there are no wall polygons on the NE wall, so try bouncing lighting off of that. You can create a loaded wand with ci("wand07",100). To see where wall polygons are or not, you can press ctrl+4 in gemrb, but ni works too. The other question is tight corridors, especially diagonal ones — if you shoot the lightning in a zigzag pattern, where are the points of contact?

**Comment by caillean7** [comment]
About the wall polygon thing: In EE, it bounces off if there are none:

![lightn1](https://user-images.githubusercontent.com/54738589/65494461-30c25600-deb5-11e9-83e3-20dd75fadca4.jpg)

Will try more with the tight corridors tonight.

**Comment by caillean7** [comment]
>  where are the points of contact?

Question on the side - do you just want the positions or also measurements of distances or the impact and recoil angles?

**Comment by lynxlynxlynx** [comment]
Coordinates are not needed if you make pics like the above - it's clear where the hit was. What will later matter is where they are in relation to the searchmap cells and/or walls.

**Comment by caillean7** [comment]
>  it looks like it bounced from a horizontal plane

Yeah, it makes those 'T shapes' all the time:
![lightn](https://user-images.githubusercontent.com/54738589/65519192-e3aaa800-dee5-11e9-9739-46b41c5bbbe1.jpg)

**Comment by MarcelHB** [comment]
Roughly an idea, did not test the lightning thoroughly recently:
- Calculate the collision with some Y offset high enough of an/the actors height, depending on the direction northwards or southwards: Just to give the correct z-impression.
- Test the search map for something blocking.
- If obstructing, take the neighbouring 8 pixels to estimate the angle of the wall. The number of possible shapes is limited but probably requires some edge cases (standalone pillar, map borders).
- Calculate the reflection.

I immediately see a case where this needs to be evaluated though (IWD2, AR1000, Targos harbor) since the western border of the map is a cliff wall that has no visible distinction in the searchmap, such as being a wall. This probably would shoot the lightning beyond the map even if physically questionable.

**Comment by lynxlynxlynx** [comment]
@caillean7 huh, that was shot in the NW direction and bounced by 90°? Looks so very wrong and again like it bounced of a horizontal. Luckily we don't have to reproduce the madness.

**Comment by caillean7** [comment]
It WAS shot in the NW direction :D

**Comment by lynxlynxlynx** [comment]
Ok, please try also some walls in other directions, preferably also some that aren't about 45°.

**Comment by caillean7** [comment]
I tried some cave walls, since most corridors have those angles. For starters here are 10 recordings, screenshots showing bounce-off points and trajectories plus the same captured on video. Let me know if you need more / different ones.

[bounce-test.zip](https://github.com/gemrb/gemrb/files/3648161/bounce-test.zip)

**Comment by lynxlynxlynx** [comment]
thanks, many look sane in there. But others muddy matters again. For example, bounce 5: at point 3/5 one would expect the reflection to go the other way.

It looks like the searchmap is the way to go. It would also explain the horizontals from before — on slanted corridors, the searchmap will be jagged, so plenty of horizontals and verticals. That could also explain bouce 5.

**Comment by caillean7** [comment]
If you look at the videos in slow motion, it looks like they're not really bouncing the bolts. To me it rather looks as if a new projectile is created from the contact point on the wall.

**Comment by lynxlynxlynx** [comment]
It has to reorient instantly, otherwise they'd have to have animations also for half-bent bolts.

**Comment by caillean7** [comment]
Ahem, yes, that makes sense.

**Comment by bradallred** [comment]
> * Calculate the collision with some ~Y~ vector offset high enough of an/the actors height, depending on the direction northwards or southwards: Just to give the correct z-impression.

I don't think it will be necessary. I suspect that the collision point is at either the center or caster end of the projectile, so by the time the collision happens its already positioned "above" it. possibly too far above it judging by the original behavior.

> * Test the search map for something blocking.
already done

> * If obstructing, take the neighbouring 8 pixels to estimate the angle of the wall. The number of possible shapes is limited but probably requires some edge cases (standalone pillar, map borders).

Sounds like my searchmap sampling :) Curious why you think pixel sampling would work better.

> I immediately see a case where this needs to be evaluated though (IWD2, AR1000, Targos harbor) since the western border of the map is a cliff wall that has no visible distinction in the searchmap, such as being a wall. This probably would shoot the lightning beyond the map even if physically questionable.

do you know how the original works here? I suspect the projectile will fly off screen. I don't see how else it could know anything about a wall if it's not in the search map.

`WallPolygons` are certainly not useful... not only are there not enough of them, but there are wall polygons for railings and all sorts of other things that shouldn't block it. I think searchmap sampling is the way to go... even though it is too course to get a great idea of slope the bright side is we should only have to take 2 additional samples :)

**Comment by bradallred** [comment]
![image](https://user-images.githubusercontent.com/1126744/65563445-d8b43f80-df06-11e9-9a76-ca737062572e.png)

This is what I have in mind. We should only need to sample searchmap nodes that are adjacent to both the projectiles current location and the wall it will hit next step.

**Comment by lynxlynxlynx** [comment]
Interesting patch 2.6 note on bugs:
> (39237) Lightning Bolt Behavior
> In the original games, if you had three creatures in a line (A, B, and C) and cast Lightning Bolt at creature B, it would pass harmlessly through A and then damage B and C. In the EEs, this was changed so that it would affect all three creatures--however, this also meant it could be dodged. After discussion, it was decided to revert this behavior to the same as the original games.

**Comment by lynxlynxlynx** [comment]
Found the logic in a recent discord chat — Bubb reversed it:

It's based on the search map.

Update projectile position based on delta - did projectile have LOS from its old position to its new position?

>     -> NO
>         -> is abs(y delta) greater than abs(x delta)?
>             -> YES
>                 -> check the position that would have resulted if the y delta was inverted - is it passible?
>                     -> YES
>                         -> revert initial y movement, invert y delta, and update projectile position based on new y delta
>                     -> NO
>                         -> revert initial x movement, invert x delta, and update projectile position based on new x delta - is the resulting position passible?
>                             -> NO
>                                 -> revert initial y movement, invert y delta, and update projectile position based on new y delta
>             -> NO
>                 -> check the position that would have resulted if the x delta was inverted - is it passible?
>                     -> YES
>                         -> revert initial x movement, invert x delta, and update projectile position based on new x delta
>                     -> NO
>                         -> revert initial y movement, invert y delta, and update projectile position based on new y delta - is the resulting position passible?
>                             -> NO
>                                 -> revert initial x movement, invert x delta, and update projectile position based on new x delta

For projectiles the search map means:

> Black        - Impassible
> Dark Red     - Passible
> Dark Green   - Passible
> Dark Yellow  - Passible
> Dark Blue    - Passible
> Dark Magenta - Passible
> Dark Cyan    - Passible
> Gray         - Passible
> Dark Gray    - Passible
> Red          - Passible
> Green        - Impassible
> Yellow       - Passible
> Blue         - Passible
> Magenta      - Impassible
> Cyan         - Passible
> White        - Passible
